### PR TITLE
feat(jax): zbl

### PR DESCRIPTION
**Pull request by njzjz**
## Release Notes

- **New Features**
	- Introduced new classes: `DPZBLLinearEnergyAtomicModel` and `PairTabAtomicModel`, enhancing atomic model functionalities.
	- Added `get_zbl_model` function for constructing `DPZBLModel` from input data.
	- Improved error handling in vector normalization with `safe_for_vector_norm` and `safe_for_sqrt`.

- **Bug Fixes**
	- Enhanced distance calculations in `format_nlist` to prevent NaN errors.

- **Documentation**
	- Updated comments and docstrings for clarity on recent changes.

- **Tests**
	- Enhanced test support for JAX backend in `test_zbl_ener.py`.

### CODE REVIEW

File: deepmd/dpmodel/model/make_model.py [hunk 1]
```diff
@@ -475,11 +475,14 @@ def _format_nlist(
...
-                rr, ret_mapping = xp.sort(rr, axis=-1), xp.argsort(rr, axis=-1)
+                # jax raises NaN error using norm
+                # but note: we don't actually need to sqrt here; the squared value is enough
+                # rr = xp.linalg.norm(coord0[:, :, None, :] - coord1, axis=-1)
+                rr2 = xp.sum(xp.square(coord0[:, :, None, :] - coord1), axis=-1)
+                rr2 = xp.where(m_real_nei, rr2, float("inf"))
+                rr2, ret_mapping = xp.sort(rr2, axis=-1), xp.argsort(rr2, axis=-1)
                 ret = xp.take_along_axis(ret, ret_mapping, axis=2)
-                ret = xp.where(rr > rcut, -1, ret)
+                ret = xp.where(rr2 > rcut * rcut, -1, ret)
```
Comment (njzjz) [inline deepmd/dpmodel/model/make_model.py:485]: I have to modify here to fix NaN forces. However, I don't quite understand why the energy has a gradient to the neighbor llist. Did I miss something?

File: deepmd/dpmodel/utils/nlist.py [hunk 2]
```diff
@@ -215,30 +215,36 @@ def build_multiple_neighbor_list(
...
+    coord2 = xp_take_along_axis(coord1, index, axis=1)
+    coord2 = xp.reshape(coord2, (nb, nloc, nsel, 3))
     diff = coord2 - coord0[:, :, None, :]
-    rr = np.linalg.norm(diff, axis=-1)
-    rr = np.where(nlist_mask, float("inf"), rr)
+    # jax raises NaN error using norm
+    # but note: we don't actually need to sqrt here; the squared value is enough
+    # rr = xp.linalg.vector_norm(diff, axis=-1)
+    rr2 = xp.sum(xp.square(diff), axis=-1)
+    rr2 = xp.where(nlist_mask, xp.full_like(rr2, float("inf")), rr2)
```
Comment (njzjz) [inline deepmd/dpmodel/utils/nlist.py:240]: This is another place that raises NaN, but I don't understand why nlist matters.

File: deepmd/dpmodel/descriptor/dpa1.py [hunk 3]
```diff
@@ -943,7 +946,7 @@ def call(
         else:
             raise NotImplementedError
 
-        normed = xp.linalg.vector_norm(
+        normed = safe_for_vector_norm(
```
Comment (njzjz) [inline deepmd/dpmodel/descriptor/dpa1.py:949]: I have to hack here

File: deepmd/dpmodel/model/make_model.py (hunk 1 above)
Comment (njzjz) [reply deepmd/dpmodel/model/make_model.py:485]: Anyway, I think avoiding sqrt also improves performance.

Comment (wanghan-iapcm) [reply deepmd/dpmodel/model/make_model.py:485]: Interesting. I agree with you.

Comment (wanghan-iapcm) [reply deepmd/dpmodel/model/make_model.py:485]: stop gradient at the nlist? https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.stop_gradient.html

**Review by wanghan-iapcm — COMMENTED** [review]
the autodiff should not pass the nlist, see my comment on using the jax.lax.stop_gradient that may fix the issue.

File: deepmd/dpmodel/model/make_model.py (hunk 1 above)
Comment (njzjz) [reply deepmd/dpmodel/model/make_model.py:485]: I found I couldn't reproduce the original error after reverting this piece of code. Not sure the reason...

Comment (njzjz) [reply deepmd/dpmodel/model/make_model.py:485]: But I think the reason of NaN is that the neighbor list is rebuilt in the linear atomic model.

Comment (njzjz) [reply deepmd/dpmodel/model/make_model.py:485]: Ok, 84803b847e641cf5a688bf2b6b9c16f3632b25c1 should fix it. I am wondering whether nlist in pytorch also has the gradient.

**Comment by wanghan-iapcm** [comment]
are you planning to revert the modification on the nlist?

**Comment by njzjz** [comment]
> are you planning to revert the modification on the nlist?

Reverted.

**Review by anyangml — APPROVED** [review]

**Review by wanghan-iapcm — APPROVED** [review]
